### ISSUE TITLE
mod_redirect problem on digitalocean install

### ISSUE THREAD

**Issue by kioinc** [bakerframework/bakercloud-ce-console#11]
I was watching your Countly installation video and you mentioned that you setup Magrocket on a DigitalOcean droplet.  Since I had previously installed Magrocket on a shared Hostgator server that I was paying $8/month for, I thought I would try to do an install on DigitalOcean, as it is cheaper to start with and scalable.

So, I setup a droplet using Ubuntu 12.04 and installed the LAMP stack.

I setup two subdomains on my main domain using VirtualHosts.  (For this install, I have \* pointed to @ with a cname.  I don't have app and api set up with their own A records or cnames.) 
The two subdomains are:
api.mydomain.com
app.mydomain.com

I was able to install the console without any problems.

I made the changes in the index.php file for the api so that it has the correct database information. 

I renamed the htaccess.txt file to .htaccess.  I didn't change any of the contents of the .htacess file.
## However, when I go to the api.mydomain and click on Check DB Connectivity, I get the following:

Error
## A website error has occured. The website administrator has been notified of the issue. Sorry for the temporary inconvenience.

However, when I go to:

http://api.mydomain.com/issues/com.nin9creative.baker/B7340733-CC87-420B-888A-EA24776E276A

I am able to connect just fine and can see the sample data.

I think that there is something wrong with the way mod_redirect is rewriting the URL, but I'm not sure. 
## Here is the redirect log file: (I changed my actual domain name to mydomain, but otherwise, this is the text of the file.)

[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](3) [perdir /var/www/api/] add path info postfix: /var/www/api/checkinstall -> /var/www/api/checkinstall/
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](3) [perdir /var/www/api/] strip per-dir prefix: /var/www/api/checkinstall/ -> checkinstall/
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](3) [perdir /var/www/api/] applying pattern '^' to uri 'checkinstall/'
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](4) [perdir /var/www/api/] RewriteCond: input='/var/www/api/checkinstall' pattern='!-f' => matched
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](2) [perdir /var/www/api/] rewrite 'checkinstall/' -> 'index.php'
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](3) [perdir /var/www/api/] add per-dir prefix: index.php -> /var/www/api/index.php
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](2) [perdir /var/www/api/] strip document_root prefix: /var/www/api/index.php -> /index.php
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e64058/initial](1) [perdir /var/www/api/] internal redirect with /index.php [INTERNAL REDIRECT]
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e602b0/initial/redir#1](3) [perdir /var/www/api/] strip per-dir prefix: /var/www/api/index.php -> index.php
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e602b0/initial/redir#1](3) [perdir /var/www/api/] applying pattern '^' to uri 'index.php'
[09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e602b0/initial/redir#1](4) [perdir /var/www/api/] RewriteCond: input='/var/www/api/index.php' pattern='!-f' => not-matched
## [09/Jan/2014:04:46:09 +0000] [api.mydomain.com/sid#b7479c78][rid#b6e602b0/initial/redir#1](1) [perdir /var/www/api/] pass through /var/www/api/index.php

I have uncommented the RewriteBase and still get the same error.  I have changed the RewriteBase to /api and /api/ and even /var/www/api and I get a 500 Internal Server Error when I do.

**Comment by nin9creative** [comment]
When I set up everything on Digital Ocean, I first installed ZPanel for managing some of the hosting of the domain and MySQL etc.  I don't recall that I had really any issues beyond getting the droplet setup and then configuring things via ZPanel.

If you look at this comment thread, does anything there help?

https://github.com/nin9creative/magrocket-backend/issues/76
https://github.com/nin9creative/magrocket-backend/issues/29

https://www.digitalocean.com/community/articles/how-to-install-and-configure-zpanel-on-an-ubuntu-12-04-vps

**Comment by kioinc** [comment]
I first followed the instructions on https://www.digitalocean.com/community/articles/initial-server-setup-with-ubuntu-12-04 including Step 5 Configure SSH.

Then I installed ZPanel using instructions from
https://www.digitalocean.com/community/articles/how-to-install-and-configure-zpanel-on-an-ubuntu-12-04-vps
and
http://docs.zpanelcp.com/?node=22

I created subdomains for Zpanel, app and api inside of ZPanel.

I was able to FTP and transfer files.

BakerCloud console was successfully installed in the app subdomain.

I changed htaccess.txt to .htaccess.

Unfortunately, I get the same error when I click on the Check Database Connectivity on the index.php page for the api.

Error

A website error has occured. The website administrator has been notified of the issue. Sorry for the temporary inconvenience.

I then went to the DNS manager for DigitalOcean and set up A and CNAMES for zpanel, api and app.  I still get the same error.

I have looked at
nin9creative/magrocket-backend#76
nin9creative/magrocket-backend#29

Unfortunately, I wasn't able to resolve the problem.

**Comment by nin9creative** [comment]
What happens if you SSH into the server and issue this command?

sudo a2enmod rewrite

You should get a message output.

**Comment by nin9creative** [comment]
Also, check your apache config file...

```
sudo nano /etc/apache2/sites-available/default

```

Then look for the following section (or something that looks like it for your specific case).

Make sure that the AllowOverride line says AllowOverride All not AllowOverride None

```
 <Directory /var/www/>
                Options Indexes FollowSymLinks MultiViews
                AllowOverride All
                Order allow,deny
                allow from all
 </Directory>
```

Save and exit the file after editing and then restart Apache.

```
sudo service apache2 restart
```

Let me know if any of that helps?

**Comment by nin9creative** [comment]
Did you have a chance to look at the info I posted above?

**Comment by kioinc** [comment]
Sorry, I've been under the weather the past few days.

sudo a2enmod rewrite gives the following output:
Module rewrite already enabled

I made the changes to /etc/apache2/sites-available/default and restarted Apache, but I still get the same error when I click on the Check DB Connectivity Link.

However,  if I manually browse to: 
http://api.yourdomain.com/issues/com.nin9creative.baker/B7340733-CC87-420B-888A-EA24776E276A

I am able to connect just fine.

It seems like the problem is with mod_rewrite.  I uncommented the RewriteBase line, but I still get the same error.

**Comment by nin9creative** [comment]
Hi, I am going to do an install on a Digital Ocean Droplet from scratch using ZPanel and see what happens.

That way if there is an issue - I can tell you what to fix.

**Comment by nin9creative** [comment]
I just ran through the install...  I didn't have to do anything beyond following the steps below.

Create a new Droplet (Ubuntu 12.04 x32)

Install ZPanel according to the following step by step guide.

https://www.digitalocean.com/community/articles/how-to-install-and-configure-zpanel-on-an-ubuntu-12-04-vps

I then setup CNAME records for ZPANEL / API / APP / WWW with the Digital Ocean DNS tool.  I had previously pointed my Godaddy DNS settings to the Digital Ocean Nameservers.

https://www.digitalocean.com/community/articles/how-to-set-up-a-host-name-with-digitalocean

![digitalocean_control_panel](https://f.cloud.github.com/assets/449883/1907240/4ffd40d2-7cc5-11e3-8377-c2b8a4cbbe9c.png)

I then logged into the zpanel with the zadmin account and the password created by the install.

I added the "bakercloudce.com" domain first.

![zpanel___zadmin_-_zpanel bakercloudce com](https://f.cloud.github.com/assets/449883/1907245/6cb108da-7cc5-11e3-9e2e-ce03ccc14efa.png)

I then added the subdomains for "api.bakercloudce.com" and "app.bakercloudce.com"

![zpanel___zadmin_-_zpanel2 bakercloudce com](https://f.cloud.github.com/assets/449883/1907251/7fc17a5e-7cc5-11e3-8f0e-41535e9d2352.png)

After that, I downloaded .zip files for the Baker Cloud Console CE (master branches) - you can use the release version as well - doesn't matter.

I created a FTP user in ZPanel - then logged in via FTP.

![bakercloudce__digital_ocean__-_public_html](https://f.cloud.github.com/assets/449883/1907285/4deab95e-7cc6-11e3-8e65-f7dae5b41611.png)

Then I uploaded the install files for the Console and installed it.  (normal install process including first setting up a mysql database as well as a db user, then making the config/database folder writeable etc...)

Then I uploaded the install files for the API.  I changed the config settings for the database in index.php.   Then I renamed the htaccess.txt file to .htaccess

It works.  Both the API and Console work out of the box without any further configuration changes in ZPanel.  I didn't have to do anything special to get it to work.

![mozilla_firefox](https://f.cloud.github.com/assets/449883/1907265/ff814bb6-7cc5-11e3-98bb-bce51c5abe78.png)

**Comment by kioinc** [comment]
Thanks for the walk through Andrew.  I know that it was time consuming to put it together and I really appreciate it.

I tried a new install, but I'm missing something, as I got the same error.

I'll give it a try again tomorrow when I can look at it with a fresh pair of eyes.

**Comment by nin9creative** [comment]
Ok

**Comment by kioinc** [comment]
I tried a fresh install again tonight, following the directions step by step, but am still getting the same error when I click on the Check DB Connectivity link.

Here is what the rewrite log gives me:

[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e44058/initial](3) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] add path info postfix: /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/checkinstall -> /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/checkinstall/
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e44058/initial](3) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] strip per-dir prefix: /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/checkinstall/ -> checkinstall/
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e44058/initial](3) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] applying pattern '^' to uri 'checkinstall/'
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e44058/initial](2) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] rewrite 'checkinstall/' -> 'index.php'
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e44058/initial](3) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] add per-dir prefix: index.php -> /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/index.php
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e44058/initial](2) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] strip document_root prefix: /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/index.php -> /index.php
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e44058/initial](1) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] internal redirect with /index.php [INTERNAL REDIRECT]
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e48c20/initial/redir#1](3) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] strip per-dir prefix: /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/index.php -> index.php
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e48c20/initial/redir#1](3) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] applying pattern '^' to uri 'index.php'
[15/Jan/2014:02:31:48 --0500] [api.mydomain.com/sid#b7489478][rid#b6e48c20/initial/redir#1](1) [perdir /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/] pass through /var/zpanel/hostdata/zadmin/public_html/api_mydomain_com/index.php

It looks like the rewrite rule doesn't end up adding checkinstall back onto api.mydomain.com/index.php.  

(I was looking at <a href="https://github.com/nin9creative/magrocket-backend/issues/76">Issue 76</a> and it seems like that is what mod_rewrite is supposed to do, but I've just started learning about regular expressions while researching this issue.)

**Comment by nin9creative** [comment]
Can you email me privately with access info for ssh to the droplet as well as zpanel login?  I'll try taking a look.

Regards,
Andrew  (andrew@nin9creative.com)

**Comment by kioinc** [comment]
I just emailed you the login info.  Thanks for your help.

**Comment by kioinc** [comment]
Andrew,

I was using the 1.1 release that I downloaded from htttp://http://bakerframework.com/bakercloudce

Deleting the line

logAnalyticMetric(AnalyticType::ApiInteraction,1,NULL,$app_id,$user_id);

from index.php solved the problem.

Thanks so much for your help resolving this!

**Comment by nin9creative** [comment]
Ok.  I updated that page now with a 1.1.1 release that has this fix in it so others won&#39;t run into the same issue!

I also tagged the release in GitHub.  Sorry for the bug!

Andrew
